Frameworks/UIKit/UIGraphicsFunctions.mm

```diff
@@ -22,72 +22,97 @@
-    }
+    std::lock_guard<std::mutex> lock(s_contextStackMutex);
 
-    CGContextRetain(context);
-    _currentCGContext[++(_currentCGContextDepth)] = context;
+    s_contextStack.emplace(context, _UIGraphicsContextTypeUser);
 }
 
 /**
  @Status Interoperable
 */
 void UIGraphicsPopContext() {
-    if (_currentCGContextDepth <= 0) {
-        assert(0);
+    std::lock_guard<std::mutex> lock(s_contextStackMutex);
+    if (s_contextStack.size() == 0) {
+        TraceError(TAG, L"UIGraphicsPopContext(): the context stack was empty.");
         return;
     }
-    CGContextRelease(_currentCGContext[_currentCGContextDepth]);
-    _currentCGContext[_currentCGContextDepth] = nullptr;
-    _currentCGContextDepth--;
+    s_contextStack.pop();
 }
 
 /**
  @Status Interoperable
 */
 CGContextRef UIGraphicsGetCurrentContext() {
-    return _currentCGContext[_currentCGContextDepth];
+    std::lock_guard<std::mutex> lock(s_contextStackMutex);
+    return s_contextStack.top().context.get();
 }
 
 /**
- @Status Caveat
- @Notes opaque parameter not supported
+ @Status Interoperable
 */
 void UIGraphicsBeginImageContextWithOptions(CGSize size, BOOL opaque, float scale) {
-    return;
-    // TODO(DH) GH#1124 Revisit when we have support for CGImage+IWICBitmap
-#if 0
+    // Locking occurs in UIGraphicsPushContext.
     if (scale == 0.0f) {
         scale = DisplayProperties::ScreenScale();
     }
-    CGContextRef newCtx = _CGBitmapContextCreateWithFormat((int)(size.width * scale), (int)(size.height * scale), _ColorARGB);
-    newCtx->scale = scale;
-    CGContextTranslateCTM(newCtx, 0.0f, size.height * scale);
-    CGContextScaleCTM(newCtx, scale, scale);
-    CGContextScaleCTM(newCtx, 1.0f, -1.0f);
 
-    UIGraphicsPushContext(newCtx);
-#endif
+    CGSize scaledSize{
+        size.width * scale, size.height * scale,
+    };
+    woc::unique_cf<CGColorSpaceRef> rgbColorSpace(CGColorSpaceCreateDeviceRGB());
+    CGContextRef context =
+        CGBitmapContextCreate(nullptr,
+                              scaledSize.width,
+                              scaledSize.height,
+                              8, // 32bpp (8 bits per component)
+                              scaledSize.width * 4, // Stride for RGBA/RGBX
+                              rgbColorSpace.get(),
+                              (opaque ? kCGImageAlphaNoneSkipLast : kCGImageAlphaPremultipliedLast) | kCGBitmapByteOrder32Little);
+
+    // Apply the UIKit ULO transform (and shadow projection transform)
+    CGContextTranslateCTM(context, 0.0f, size.height * scale);
+    CGContextScaleCTM(context, scale, scale);
+    CGContextScaleCTM(context, 1.0f, -1.0f);
+    _CGContextSetShadowProjectionTransform(context, CGAffineTransformMakeScale(1.0, -1.0));
+
+    {
+        std::lock_guard<std::mutex> lock(s_contextStackMutex);
+
+        s_contextStack.emplace(context, _UIGraphicsContextTypeImage, size, scale);
+    }
+
+    // This context is owned by the context stack.
+    CGContextRelease(context);
```

Comment:
keep context in a unique_cf? nothing should fail terribly in here but would keep consistent with usage